### PR TITLE
Button and Link: Adding imperative focus behavior and tests to prototypes

### PR DESCRIPTION
**Pull request by khmakoto**
This PR adds imperative focus behavior and handlers to both the `Button` and `Link` components, as well as tests to ensure they work correctly.

As always, this is an early version and any feedback is appreciated.

### CODE REVIEW

**Comment by khmakoto** [comment]
One last thing, I made a comment in the `*.test.tsx` files about having to move the enzyme adapter configuration somewhere more global. @kenotron any ideas on where to put this?

**Comment by kenotron** [comment]
I think we should start out putting the test next to the components, we could also put them in a tests folder

**Comment by layershifter** [comment]
As I understand the goal we need to have an access to some DOM methods, but from my personal experience this thing will be endless as then we will need `.blur()`, `getClientBoundingRect()`, etc.

So my question is, why don't use [`forwardRef`](https://reactjs.org/docs/forwarding-refs.html)? 🤔 Especially when other libraries do this as it drastically simplifies composition.

- https://gist.github.com/ryanflorence/e5c794e6093d16a69fa88d2112a292f7#random-implementation-checklist
- https://material-ui.com/guides/composition/#caveat-with-refs
- https://www.styled-components.com/docs/advanced#refs

File: packages/base/src/components/Button/Button.test.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import { configure, mount } from 'enzyme';
+import { ButtonBase } from './Button.base';
+import { IButton } from './Button.types';
+
+// TODO: Need to move this to some global configuration.
+const Adapter = require('enzyme-adapter-react-16');
+configure({ adapter: new Adapter() });
```
Comment (aneeshack4) [inline packages/base/src/components/Button/Button.test.tsx]: It looks like this is why the build is failing? `Enzyme Internal Error: configured enzyme adapter did not inherit from the EnzymeAdapter base class`

File: packages/base/src/components/Button/useButton.ts [hunk 2]
```diff
@@ -8,9 +8,15 @@ export interface IButtonState {
...
+  const { componentRef, disabled, onClick } = userProps;
 
-  const rootRef = React.useRef<HTMLElement>(null);
+  const rootRef = useRef<HTMLElement>(null);
+
+  useImperativeHandle(componentRef, () => ({
+    focus: () => {
+      rootRef.current && rootRef.current.focus();
+    }
+  }));
```
Comment (aneeshack4) [inline packages/base/src/components/Button/useButton.ts]: Shouldn't `forwardRef` also be used here? https://reactjs.org/docs/hooks-reference.html#useimperativehandle

File: packages/base/src/components/Slider/Slider.types.ts [hunk 3]
```diff
@@ -19,7 +19,7 @@ export interface ISliderClasses extends IClasses<ISliderSlots> {
   rootDisabled: string;
 }
 
-export interface ISliderProps extends ISlottableProps<ISliderSlotProps, ISliderClasses> {
```
Comment (aneeshack4) [inline packages/base/src/components/Slider/Slider.types.ts]: How come we're not using `ISliderSlotProps` anymore?

File: packages/base/src/components/Button/Button.test.tsx (hunk 1 above)
Comment (khmakoto) [reply packages/base/src/components/Button/Button.test.tsx]: Yeah, I'll look into it.

File: packages/base/src/components/Slider/Slider.types.ts (hunk 3 above)
Comment (khmakoto) [reply packages/base/src/components/Slider/Slider.types.ts]: I noticed that we were doing something weird in the `*.base.tsx` files when casting `slots` as `I{Component}Slots`. We shouldn't need to do that. So I changed the helper types to correctly reflect this, you can see this below by comparing the changes between `ISlots.ts` and `Slots.types.ts`.

**Comment by khmakoto** [comment]
@layershifter r @aneeshack4 I've changed the PR to use `forwardRef`, please take another look.

File: packages/base/src/components/Button/Button.test.tsx (hunk 1 above)
Comment (khmakoto) [reply packages/base/src/components/Button/Button.test.tsx]: This has been resolved.

File: packages/base/src/components/Button/useButton.ts (hunk 2 above)
Comment (khmakoto) [reply packages/base/src/components/Button/useButton.ts]: Done! Check the latest changes.

**Comment by khmakoto** [comment]
@layershifter just to add some context here, yesterday we talked about this during our team status meeting and we've decided to use `forwardRef` in scenarios like this, where we're dealing with primitives that have a 1:1 reference relationship with a DOM element inside their render function. For more ambiguous things, like focusing on a list, I think we'd still prefer the other approach I was proposing at the beginning of this PR.

Please let me know your thoughts on this and any concerns you'd have with that. Thanks!

**Pull request merged**